File: server/src/main/scala/producer/producer_session/producer_session_task/ProducerSessionTask.scala
package producer.producer_session.producer_session_task

import com.tools.teal.pulsar.ui.producer.v1.producer as pb
import _root_.producer.producer_task.ProducerTask

case class ProducerSessionTask(
    task: ProducerTask,
)

object ProducerSessionTask:
    def fromPb(v: pb.ProducerSessionTask): ProducerSessionTask =
        val task = v.task match
            case pb.ProducerSessionTask.Task.TaskProducer(v) =>
                ProducerTask.fromPb(v)
            case _ => throw Exception("Invalid producer session task type")

        ProducerSessionTask(
            task = task,
        )

    def toPb(v: ProducerSessionTask): pb.ProducerSessionTask =
        val task = v.task match
            case v: ProducerTask =>
                pb.ProducerSessionTask.Task.TaskProducer(ProducerTask.toPb(v))
            case _ => throw Exception("Invalid producer session task type")

        pb.ProducerSessionTask(
            task = task,
        )

File: ui/components/ui/ProducerSession/producer-session-config/producer-session-task/producer-session-task.ts
import * as pb from '../../../../../grpc-web/tools/teal/pulsar/ui/producer/v1/producer_pb';
import { ProducerTask, producerTaskFromPb, producerTaskToPb } from '../../producer-task/producer-task';

export type ProducerSessionTask = {
  type: 'producer-session-task',
  task: { type: 'producer-task', task: ProducerTask }
};

export function producerSessionTaskFromPb(v: pb.ProducerSessionTask): ProducerSessionTask {
  let task: ProducerSessionTask['task'];

  switch (v.getTaskCase()) {
    case pb.ProducerSessionTask.TaskCase.TASK_PRODUCER:
      task = { type: 'producer-task', task: producerTaskFromPb(v.getTaskProducer()!) };
      break;
    default:
      throw new Error(`Unknown task case: ${v.getTaskCase()}`);
  }

  return {
    type: 'producer-session-task',
    task
  };
}

export function producerSessionTaskToPb(v: ProducerSessionTask): pb.ProducerSessionTask {
  const producerSessionTaskPb = new pb.ProducerSessionTask();

  switch (v.task.type) {
    case 'producer-task':
      producerSessionTaskPb.setTaskProducer(producerTaskToPb(v.task.task));
      break;
    default:
      throw new Error(`Unknown task type: ${v.task.type}`);
  }

  return producerSessionTaskPb;
}
